### PR TITLE
Shooter

### PR DESCRIPTION
**Pull request by greenblitz4590**

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/MotorFlyWheel/FalconFlyWheel/FalconFlyWheel.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	}
+	
+	@Override
+	public void setVoltage(double voltage) {
+		motor.setVoltage(voltage);
+	}
+	
+	@Override
+	public void setVelocity(double velocity) {
+		motor.setControl(new MotionMagicVelocityDutyCycle(
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/MotorFlyWheel/FalconFlyWheel/FalconFlyWheel.java:31]: motion magic is not the correct use for it at that case.
the MotionMagicVelocityDutyCycle will generate a profile of velocity every cycle of the motor and this action might slow him down. for this case I recommend using VelocityVoltage

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/MotorFlyWheel/MotorFlyWheelConstants.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import com.ctre.phoenix6.configs.ClosedLoopRampsConfigs;
+import com.ctre.phoenix6.configs.CurrentLimitsConfigs;
+import com.ctre.phoenix6.configs.MotionMagicConfigs;
+import com.ctre.phoenix6.signals.NeutralModeValue;
+import edu.wpi.first.math.controller.SimpleMotorFeedforward;
+import edu.wpi.first.math.interpolation.InterpolatingTreeMap;
+import edu.wpi.first.math.interpolation.Interpolator;
+import edu.wpi.first.math.interpolation.InverseInterpolator;
+
+public class MotorFlyWheelConstants extends edu.greenblitz.robotName.subsystems.shooter.FlyWheel.FlyWheelConstants {
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/MotorFlyWheel/MotorFlyWheelConstants.java:12]: why does the constants class extends another class?-

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/MotorFlyWheel/neoFlyWheel/NeoFlyWheel.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                velocity,
+                CANSparkMax.ControlType.kVelocity,
+                MotorFlyWheelConstants.PID_SLOT,
+                MotorFlyWheelConstants.FEEDFORWARD.calculate(velocity)
+        );
+    }
+
+    @Override
+    public void updateInputs(FlyWheelInputsAutoLogged inputs) {
+        inputs.appliedOutput = motor.getAppliedOutput();
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/MotorFlyWheel/neoFlyWheel/NeoFlyWheel.java:41]: multyply by battery voltage

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/simulationFlyWheel/SimulationFlyWheel.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                voltage,
+                RobotConstants.SimulationConstants.MIN_MOTOR_VOLTAGE,
+                RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE
+        );
+        flyWheelSimulation.setInputVoltage(appliedVoltage);
+    }
+
+    @Override
+    public void setVelocity(double velocity) {
+        //todo add
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/simulationFlyWheel/SimulationFlyWheel.java:39]: why todo? just add the PIDController

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheelInputs.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package edu.greenblitz.robotName.subsystems.shooter.FlyWheel;
+
+import org.littletonrobotics.junction.AutoLog;
+
+@AutoLog
+public class FlyWheelInputs {
+    public double appliedOutput;
+    public double outputCurrent;
+    public double velocity;
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheelInputs.java:9]: moreeeee.
temperate and more

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/OI.java [hunk 6]
```diff
@@ -32,7 +36,35 @@ public SmartJoystick getSecondJoystick() {
...
+                    @Override
+                    public void execute() {
+                        motor.set(-0.2);
+                    }
+
+                    @Override
+                    public void end(boolean interrupted) {
+                        motor.set(0);
+                    }
+                }
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/OI.java:67]: dont put this in master<3

File: src/main/java/edu/greenblitz/robotName/Robot.java [hunk 7]
```diff
@@ -25,18 +23,21 @@ public enum RobotType {
...
     public void robotInit() {
-        initializeLogger();
+//        initializeLogger();
 
-        SwerveChassis.init();
-        SwerveChassis.getInstance().setDefaultCommand(new MoveByJoysticks(MoveByJoysticks.DriveMode.NORMAL));
-        SwerveChassis.getInstance().resetAllEncoders();
+//        SwerveChassis.init();
+//        SwerveChassis.getInstance().setDefaultCommand(new MoveByJoysticks(MoveByJoysticks.DriveMode.NORMAL));
+//        SwerveChassis.getInstance().resetAllEncoders();
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/Robot.java:30]: give them back

File: src/main/java/edu/greenblitz/robotName/Robot.java [hunk 8]
```diff
@@ -25,18 +23,21 @@ public enum RobotType {
...
 
         OI.getInstance();
         CommandScheduler.getInstance().enable();
     }
     @Override
     public void robotPeriodic() {
         CommandScheduler.getInstance().run();
+        SmartDashboard.putNumber("velocity", FlyWheel.getInstance().getVelocity());
+        SmartDashboard.putNumber("current", FlyWheel.getInstance().getCurrent());
+        SmartDashboard.putNumber("voltage", FlyWheel.getInstance().getVoltage());
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/Robot.java:40]: not in the periodic

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package edu.greenblitz.robotName.commands.shooter;
+
+import edu.greenblitz.robotName.subsystems.shooter.FlyWheel.FlyWheelConstants;
+
+public class ShootByVelocity extends FlyWheelCommand {
+
+    private int isAtShootingSpeed;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java:7]:
```suggestion
    private int timeInShootingSpeed;
```

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheel.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    private final FlyWheelInputsAutoLogged flyWheelInputs;
+    private boolean isPreparedToShoot;
+
+    private FlyWheel() {
+        flyWheel = FlyWheelFactory.create();
+        flyWheelInputs = new FlyWheelInputsAutoLogged();
+        flyWheel.updateInputs(flyWheelInputs);
+    }
+
+    public static FlyWheel getInstance() {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheel.java:18]: add an init function

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheelConstants.java [hunk 11]
```diff
@@ -0,0 +1,6 @@
+package edu.greenblitz.robotName.subsystems.shooter.FlyWheel;
+
+public class FlyWheelConstants {
+    public static final double EPSILON_RPM = 50;
+    public static final double SHOOTING_SPEED_TIME = 1;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheelConstants.java:5]:
```suggestion
    public static final double TIME_UNTIL_STABLE = 1;
```

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheelFactory.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import edu.greenblitz.robotName.RobotConstants;
+import edu.greenblitz.robotName.subsystems.shooter.FlyWheel.MotorFlyWheel.neoFlyWheel.NeoFlyWheel;
+import edu.greenblitz.robotName.subsystems.shooter.FlyWheel.simulationFlyWheel.SimulationFlyWheel;
+
+public class FlyWheelFactory {
+    public static IFlyWheel create() {
+        return switch (RobotConstants.ROBOT_TYPE) {
+            case ROBOT_NAME -> new NeoFlyWheel();
+            case REPLAY -> new ReplayFlyWheel();
+            default -> new SimulationFlyWheel();
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheelFactory.java:12]: add a simulation case

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+public class ShootByVelocity extends FlyWheelCommand {
+
+    private int timeInShootingSpeed;
+    private double velocity;
+
+    public ShootByVelocity(double velocity) {
+        this.velocity = velocity;
+    }
+
+    public boolean isFlywheelAtVelocity() {
```
Comment (itamaroryan) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java:14]: Why is this public?

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public void initialize() {
+        timeInShootingSpeed = 0;
+    }
+
+    @Override
+    public void execute() {
+        flyWheelCommand.setVelocity(velocity);
```
Comment (itamaroryan) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java:25]: Why is the class called flyWheelCommand?

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        if (isFlywheelAtVelocity()) {
+            timeInShootingSpeed++;
+        } else {
+            timeInShootingSpeed = 0;
+        }
+        flyWheelCommand.setPreparedToShoot(timeInShootingSpeed > FlyWheelConstants.SHOOTING_SPEED_TIME);
```
Comment (itamaroryan) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java:31]: Why is the class called flyWheelCommand?

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FalconFlyWheel/FalconFlyWheel.java
Comment (itamaroryan) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FalconFlyWheel/FalconFlyWheel.java:1]: Ctrl+Alt+L this file <3

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheel.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    public void setPreparedToShoot(boolean isPreparedToShoot) {
+        this.isPreparedToShoot = isPreparedToShoot;
+    }
+
+    public double getVelocity() {
+        return flyWheelInputs.velocity;
+    }
+    public double getCurrent (){
```
Comment (itamaroryan) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/FlyWheel.java:48]: Why is there a white space here :(

**Review by itamaroryan — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public void execute() {
+        flyWheel.setVelocity(velocity);
+        if (isFlywheelAtVelocity()) {
+            timeInShootingSpeed++;
+        } else {
+            timeInShootingSpeed = 0;
+        }
+        flyWheel.setPreparedToShoot(timeInShootingSpeed > FlyWheelConstants.SHOOTING_SPEED_TIME);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java:31]: put in isFinished

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    private int timeInShootingSpeed;
+    private double velocity;
+
+    public ShootByVelocity(double velocity) {
+        this.velocity = velocity;
+    }
+
+    private boolean isFlywheelAtVelocity() {
+        return Math.abs(flyWheel.getVelocity() - velocity) < FlyWheelConstants.EPSILON_RPM;
+    }
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootByVelocity.java:16]: not supposed to be in the command

File: src/main/java/edu/greenblitz/robotName/subsystems/Battery.java [hunk 19]
```diff
@@ -9,11 +9,9 @@
 
 public class Battery extends GBSubsystem {
 	
-	private static final double MIN_VOLTAGE_BATTERY = 11.97;
+	private static final double MIN_VOLTAGE_BATTERY = 0;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/Battery.java:12]: dont merge this

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/simulationFlyWheel/SimulationFlyWheel.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+import edu.greenblitz.robotName.subsystems.shooter.FlyWheel.IFlyWheel;
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.simulation.FlywheelSim;
+
+public class SimulationFlyWheel implements IFlyWheel {
+    FlywheelSim flyWheelSimulation;
+    private double appliedVoltage;
+    PIDController pidController;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/FlyWheel/simulationFlyWheel/SimulationFlyWheel.java:14]:
```suggestion
    private FlywheelSim flyWheelSimulation;
    private double appliedVoltage;
    private PIDController pidController;
```

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Modules/SwerveModuleFactory.java [hunk 21]
```diff
@@ -14,7 +14,7 @@ public static ISwerveModule create(SwerveChassis.Module module) {
 
         return switch (RobotConstants.ROBOT_TYPE) {
             case REPLAY -> new ReplaySwerveModule();
-            case ROBOT_NAME -> new MK4ISwerveModule(module);
+            case ROBOT_NAME -> new KazaSwerveModule(module);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Modules/SwerveModuleFactory.java:17]: give the sds back

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

**Review by itamaroryan — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Pull request merged**